client/src/main/java/io/pravega/client/stream/impl/SegmentSelector.java

```diff
@@ -64,24 +74,26 @@
      *
      * @param routingKey The key that should be used to select from the segment that the event
      *            should go to.
+     * @param length The length of the data being written.
      * @return The SegmentOutputStream for the segment that has been selected or null if
      *         {@link #refreshSegmentEventWriters(Consumer)} needs to be called.
      */
     @Synchronized
-    public SegmentOutputStream getSegmentOutputStreamForKey(String routingKey) {
+    public SegmentOutputStream getSegmentOutputStreamForKey(String routingKey, int length) {
         if (currentSegments == null) {
             return null;
         }
-        return writers.get(getSegmentForEvent(routingKey));
+        return writers.get(getSegmentForEvent(routingKey, length));
     }
 
     @Synchronized
-    public Segment getSegmentForEvent(String routingKey) {
+    public Segment getSegmentForEvent(String routingKey, int length) {
         if (currentSegments == null) {
             return null;
         }
         if (routingKey == null) {
-            return currentSegments.getSegmentForKey(random.nextDouble());
+            bytesWritten += length;
+            return currentSegments.getSegmentForKey(randomRoutingKeyHasher.hashToRange(bytesWritten / BYTES_PER_SEGMENT));
```

Comment:
This would return the same routing key for every 1GB of data, right ? (2^12 + BYTES_PER_SEGMENT)
(first 1GB of data with random routing key will be written to segment corresponding to routing key as 0.0)

Reply:
I am not sure what you are getting at. bytesWritten is a long, so it should only have the same value every 2^64 bytes written. By dividing we are throwing out the low order bits so that the value hashed only changes every 256kb.

Reply:
oops i misread the code. Yeah the routing key changes every `262144` bytes.